### PR TITLE
Update reduce_scatter_tests to work for world_size > 1

### PR DESCRIPTION
**Pull request by H-Huang**
These tests only worked since `world_size==1`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104424

### CODE REVIEW

**Review by kwen2501 — CHANGES_REQUESTED** [review]

File: test/distributed/test_c10d_nccl.py [hunk 1]
```diff
@@ -2990,6 +2990,39 @@ def test_consecutive_comm_work_wait_gpu(self):
...
+    @skip_if_lt_x_gpu(2)
+    def test_reduce_scatter_tensor_coalesced(self):
+        store = dist.FileStore(self.file_name, self.world_size)
+        dist.init_process_group(
+            "nccl",
+            world_size=self.world_size,
+            rank=self.rank,
+            store=store,
+        )
+        output_tensors = torch.zeros(2, 2).to(self.rank)
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply test/distributed/test_c10d_nccl.py:3019]: I'm not entirely sure how coalesced collectives work, is it able to operate on different parts of the same tensor? In this test world_size=2 so

output_tensors = [[0, 0], [0, 0]]
input_tensors = [[[1, 1],[1, 1]], [[1, 1],[1, 1]]]

coalesced (twice)
dist.reduce_scatter_tensor([0,0], [[1,1], [1,1]])

so each row of output_tensors ([0,0]) will be updated with reduce_scatter from itself and the other rank, for a total of 4 reduce_scatters. The test is passing, but does that assumption work?

**Comment by H-Huang** [comment]
CI failures are due to Github outage, will rerun when github is back

File: test/distributed/test_c10d_nccl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/test_c10d_nccl.py:3019]: @H-Huang  this matches my understanding. 

@kwen2501 The test looks okay to me: @H-Huang  is using dim-0 of `output_tensors` as the "list".

**Review by awgu — APPROVED** [review]
Thanks!
